aws_signing_helper/tpm_signer.go

```diff
@@ -0,0 +1,345 @@
+		if err != nil {
+			return nil, err
+		}
+	}
+	return signature, nil
+}
+
+// Gets the x509.Certificate associated with this TPMv2Signer
+func (tpmv2Signer *TPMv2Signer) Certificate() (*x509.Certificate, error) {
+	return tpmv2Signer.cert, nil
+}
+
+// Gets the certificate chain associated with this TPMv2Signer
+func (tpmv2Signer *TPMv2Signer) CertificateChain() (chain []*x509.Certificate, err error) {
+	return tpmv2Signer.certChain, nil
+}
+
+/*
+ * DER forbids storing a BOOLEAN as anything but 0x00 or 0xFF,
+ * 0x01, and the Go asn1 parser cannot be relaxed. But both
+ * OpenSSL ENGINEs which produce these keys have at least in
+ * the past emitted 0x01 as the value, leading to an Unmarshal
+ * failure with 'asn1: syntax error: invalid boolean'. So...
+ */
+func fixupEmptyAuth(tpmData *[]byte) {
+	var pos int = 0
+
+	// Skip the SEQUENCE tag and length
+	if len(*tpmData) < 2 || (*tpmData)[0] != 0x30 {
+		return
+	}
+
+	// Don't care what the SEQUENCE length is, just skip it
+	pos = 1
+	lenByte := (*tpmData)[pos]
+	if lenByte < 0x80 {
+		pos = pos + 1
+	} else if lenByte < 0x85 {
+		pos = pos + 1 + int(lenByte) - 0x80
+	} else {
+		return
+	}
+
+	if len(*tpmData) <= pos {
+		return
+	}
+
+	// Use asn1.Unmarshal to eat the OID; we care about 'rest'
+	var oid asn1.ObjectIdentifier
+	rest, err := asn1.Unmarshal((*tpmData)[pos:], &oid)
+	if err != nil || rest == nil || !oid.Equal(oidLoadableKey) || len(rest) < 5 {
+		return
+	}
+
+	// If the OPTIONAL EXPLICIT BOOLEAN [0] exists, it'll be here
+	pos = len(*tpmData) - len(rest)
+
+	if (*tpmData)[pos] == 0xa0 && // Tag
+		(*tpmData)[pos+1] == 0x03 && // length
+		(*tpmData)[pos+2] == 0x01 &&
+		(*tpmData)[pos+3] == 0x01 &&
+		(*tpmData)[pos+4] == 0x01 {
+		(*tpmData)[pos+4] = 0xff
+	}
+}
+
+// Returns a TPMv2Signer, that can be used to sign a payload through a TPMv2-compatible
+// cryptographic device
+func GetTPMv2Signer(certificate *x509.Certificate, certificateChain []*x509.Certificate, keyPem *pem.Block) (signer Signer, signingAlgorithm string, err error) {
+
+	var tpmData tpm2_TPMKey
+
+	fixupEmptyAuth(&keyPem.Bytes)
+	_, err = asn1.Unmarshal(keyPem.Bytes, &tpmData)
+	if err != nil {
+		return nil, "", err
+	}
+
+	if !tpmData.Oid.Equal(oidLoadableKey) {
+		return nil, "", errors.New("Invalid OID for TPMv2 key:" + tpmData.Oid.String())
+	}
+
+	if tpmData.Policy != nil || tpmData.AuthPolicy != nil {
+		return nil, "", errors.New("TPMv2 policy not implemented yet")
+	}
+	if tpmData.Secret != nil {
+		return nil, "", errors.New("TPMv2 key has 'secret' field which should not be set")
+	}
+
+	if !handleIsPersistent(tpmData.Parent) &&
+		tpmData.Parent != int(tpm2.HandleOwner) &&
+		tpmData.Parent != int(tpm2.HandleNull) &&
+		tpmData.Parent != int(tpm2.HandleEndorsement) &&
+		tpmData.Parent != int(tpm2.HandlePlatform) {
+		return nil, "", errors.New("Invalid parent for TPMv2 key")
+	}
+	if len(tpmData.Pubkey) < 2 ||
```

Comment:
We have to do all this validation ourselves? This and the functionality encapsulated within the `fixupEmptyAuth` function feel like things that should be done by `go-tpm`. I haven't done a deep dive on what functionality is currently supported by the repository though (and I do see that it says that the entire TPM 2.0 specification isn't yet supported). This just feels like something we shouldn't have to do ourselves.

Reply:
Agreed. https://github.com/google/go-tpm/issues/312